stdlib/seq.mli

```diff
@@ -15,95 +15,726 @@
-    or [fun () -> Cons (x, unfold f y)] if [f u] returns [Some (x, y)].
+    [zip xs ys] is equivalent to [map2 (fun a b -> (a, b)) xs ys].
+
+    @since 4.14 *)
+
+val map2 : ('a -> 'b -> 'c) -> 'a t -> 'b t -> 'c t
+(** [map2 f xs ys] is the sequence of the elements [f x y],
+    where the pairs [(x, y)] are drawn synchronously from the
+    sequences [xs] and [ys].
+
+    If the sequences [xs] and [ys] have different lengths, then
+    the sequence ends as soon as one sequence is exhausted;
+    the excess elements in the other sequence are ignored.
+
+    [map2 f xs ys] is equivalent to [map (fun (x, y) -> f x y) (zip xs ys)].
+
+    @since 4.14 *)
+
+val interleave : 'a t -> 'a t -> 'a t
+(** [interleave xs ys] is the sequence that begins with the first element of
+    [xs], continues with the first element of [ys], and so on.
+
+    When one of the sequences [xs] and [ys] is exhausted,
+    [interleave xs ys] continues with the rest of the other sequence.
+
+    @since 4.14 *)
+
+val sorted_merge : ('a -> 'a -> int) -> 'a t -> 'a t -> 'a t
+(** If the sequences [xs] and [ys] are sorted according to the total preorder
+    [cmp], then [sorted_merge cmp xs ys] is the sorted sequence obtained by
+    merging the sequences [xs] and [ys].
+
+    For more details on comparison functions, see {!Array.sort}.
+
+    @since 4.14 *)
+
+val product : 'a t -> 'b t -> ('a * 'b) t
+(** [product xs ys] is the Cartesian product of the sequences [xs] and [ys].
+
+    For every element [x] of [xs] and for every element [y] of [ys],
+    the pair [(x, y)] appears once as an element of [product xs ys].
+
+    The order in which the pairs appear is unspecified.
+
+    The sequences [xs] and [ys] are not required to be finite.
+
+    The sequences [xs] and [ys] must be persistent.
+
+    @since 4.14 *)
+
+val product_with : ('a -> 'b -> 'c) -> 'a t -> 'b t -> 'c t
+(** The sequence [product_with f xs ys] is the image through [f] of the Cartesian
+    product of the sequences [xs] and [ys].
+
+    For every element [x] of [xs] and for every element [y] of [ys],
+    the element [f x y] appears once as an element of [product_with f xs ys].
+
+    The order in which these elements appear is unspecified.
+
+    The sequences [xs] and [ys] are not required to be finite.
+
+    The sequences [xs] and [ys] must be persistent.
+
+    [product_with f xs ys] is equivalent to
+    [map (fun (x, y) -> f x y) (product xs ys)].
+
+    @since 4.14 *)
+
+(** {1 Splitting a sequence into two sequences} *)
+
+val unzip : ('a * 'b) t -> 'a t * 'b t
+(** [unzip] transforms a sequence of pairs into a pair of sequences.
+
+    [unzip xs] is equivalent to [(map fst xs, map snd xs)].
+
+    Querying either of the sequences returned by [unzip xs]
+    causes [xs] to be queried.
+    Therefore, querying both of them
+    causes [xs] to be queried twice.
+    Thus, [xs] must be persistent and cheap.
+    If that is not the case, use [unzip (memoize xs)].
```

Comment:
If `xs` is ephemeral but a long sequence, then `unzip (memoize xs)` will memoize a large number of elements and might be expensive. It might be interesting to have a function `unzip_ephemeral xs` which will memoize only the computed but not yet queried elements of both result sequences, so that if the use of the two sequences are almost balanced, memory footprint remains minimal (and likewise with `partition_either` and other functions below).

Reply:
If I understand what you mean, some kind of mutable buffer is needed. I have implemented this at one point but it is complex and ugly. I don't think we want this is the standard library.

Reply:
> If xs is ephemeral but a long sequence, then unzip (memoize xs) will memoize a large number of elements and might be expensive.

Wait, is this correct? I would optimistically expect this function to memoize elements of `xs` on-demand (thunks are only created for the prefix corresponding to the longest-forced-prefix of the two output sequences), and the GC to drop elements once they become unreachable from both sides. At a given point in time, the live thunks should only correspond to the difference between the two result sequences.

(Well, I *know* that memoization is on-demand, it's in the code, and I *think* that elements become dead quickly enough, but that's more optimistic and a test would clarify that.)

Reply:
If you write `let (ys, zs) = unzip (memoize xs)` then the elements of `xs` are memoized on demand (that is, when `ys` or `zs` is consumed) and they are held in memory as long as they remain reachable. The amount of memory that is held can become large if the user consumes `ys` much faster than `zs`, or vice-versa. I could add a comment along these lines, if desired.

Reply:
(Probably my reply to Naëla was misleading. I do not clearly remember why I experimented with a mutable buffer... giving the user a choice between `unzip` and `unzip . memoize` should be fine, I can't think why one might need something else.)

Reply:
Ah yes indeed. I was naïvely thinking that all elements were held in memory once memoized, but I had forgotten about the GC. In that case, the only difference with a specialized `unzip_ephemeral` function would be that already-used elements from one sequence would not be collected if the corresponding element from the other sequence has not yet been queried, which should in most cases only lead to a constant factor increase in memory usage (but in the worst-case could be far worse).

Reply:
Right. Out of curiosity I tried to find a reasonably simple implementation of this `unzip_ephemeral` version, and I found the following:

```ocaml
let unzip (type a b) (xys : (a * b) Seq.t) =
    let xq = Queue.create () in
    let yq = Queue.create () in
    let xys = ref xys in
    let fetch () =
      match !xys () with
      | Nil -> false
      | Cons ((x, y), rest) ->
        Queue.push x xq;
        Queue.push y yq;
        xys := rest;
        true
    in
    let rec get_from q =
      match Queue.take_opt q with
      | Some v -> Some (v, q)
      | None ->
        if fetch () then get_from q
        else None
    in
    Seq.unfold get_from xq,
    Seq.unfold get_from yq
```

Note: some operations use `unzip` internally, and it may be that this implementation is better there. (I'm not sure but, if I read `transpose` correctly, it looks like it will retain the whole matrix even after the first 2 elements (for example) have been dropped from the result, when it needs not retain the first 2 elements of each row anymore.)

Reply:
Thanks for implementing it! Just in case this implementation ends up somewhere, I think in the first case of `fetch`, you would need to add a `xys := (fun () -> Nil)` so that the `xys` isn't called multiple times at the end.